working/control-flow-collections/feature-specification.md

```diff
@@ -684,10 +684,137 @@ shows that a variable has some type and promotion isn't otherwise aborted.
 
 ### Const collections
 
-An `if` element can be used in a const collection, provided its condition
-expression and body elements are all const.
+We have to be careful to ensure that arbitrary computation doesn't happen due
+to `if` or `for` appearing in a constant collection. To that end, we only allow an element in a const collection when:
 
-`for` elements cannot be used in const lists and maps.
+*   It is an expression element and the expression is constant.
+
+*   It is an `if` element and the condition expression is constant and both body
+    elements are allowed.
+
+*   It is a `for` element with an `in` clause. The iterator expression must be
+    a constant expression. In a constant list literal, the expression must
+    evaluate to an object created by a constant list or set literal, as in:
+
+    ```dart
+    const list = [1, 2];
+    const set = {3, 4};
+    const result = const [
+      for (var i in list) i,
+      for (var i in set) i
+    ];
+    ```
+
+    This restriction ensures that iterating over the iterable does not call
+    user code.
+
+
+A collection literal is now a series of *elements* (some of which may contain
+nested subelements) instead of just expressions (for lists and sets) or entries
+(for maps). A constant collection takes that tree of elements and *expands* it
+to a series of values (lists and sets) or entries (maps). The resulting
+collection contains that series of values/entries, in order.
+
+There are three kinds of elements to consider:
+
+*   An **expression element** (the base case in lists and sets):
+
+    *   It is a compile-time error if the expression is not a constant
+        expression.
+
+    The expansion is the value of the expression.
+
+*   An **entry element** (the base case in maps):
+
+    *   It is a compile-time error if the key and value expressions are not
+        constant expressions.
+
+    The expansion is the entry formed by the key and value expression values.
+
+*   An **if element**:
+
+    *   It is a compile-time error if the condition expression is not constant
+        or does not evaluate to `true` or `false`.
+
+    The expansion is:
+
+    *   The then element if the condition expression evaluates to `true`.
```

Comment:
May want to be clear that evaluation short circuits (as we are changing ?: to do).  So constant evaluation of the then/else branch doesn't happen unless the condition evaluates to true/false (respectively).